tests/lwip_sock_dns: add test

HendrikVE opened the pull request:
### Contribution description

This PR adds a test for `sock_dns` when using `lwip`. The shell interface is quite similiar to the `gnrc` version, but we are also able to use IPv4 and the test requires a real connection to the internet.

### Testing procedure

See README for test instructions.

yarrick commented on tests/lwip_sock_dns/main.c:
```diff
@@ -0,0 +1,190 @@
+static int _dns(int argc, char **argv);
+
+static const shell_command_t _shell_commands[] = {
+    { "dns", "configures and requests a DNS server", _dns },
```
requests -> queries

yarrick commented on tests/lwip_sock_dns/tests/01-run.py:
```diff
@@ -0,0 +1,66 @@
+from testrunner import run
+
+
+def _test_dns_requests(child):
```
Can we expect these results to be stable?

yarrick replied:
Maybe we should just check that the IP version is the expected one and then accept any valid address.

HendrikVE replied:
I also thought about the resolved address of `example.com` and I really don't know about their policy to keep the same IP forever. But we could check for `dns.google.com` instead. I'm pretty sure Google won't give away their nice IPs and the service is also running since 2010. Of course there will be two possible results for IPv4 and IPv6 in this case.
As it goes for the reachability I don't see that this is a huge problem unless you sit behind a great firewall or do I miss something?

yarrick commented on tests/lwip_sock_dns/main.c:
```diff
@@ -0,0 +1,190 @@
+    if ((argc > 1) && (strcmp(argv[1], "server") == 0)) {
+        return _dns_server(argc, argv);
+    }
+    else if ((argc > 2) && (strcmp(argv[1], "request") == 0)) {
```
Minor: Consider calling it `lookup` or `resolve` instead of request.

HendrikVE replied:
I'll change `tests/gnrc_sock_dns` then too as it has exactly the same terminology.

yarrick reviewed (commented):
Looks good, test passes on my ESP32 board.

miri64 reviewed (commented):
IMHO, tests should as little as possible rely on outside factors. A tester will just run the script and if it is not successful because the server is not reachable or the addresses have changed, mark the test as failed => you need to fix it.

yarrick commented on tests/lwip_sock_dns/tests/01-run.py:
```diff
@@ -0,0 +1,66 @@
+
+
+def test_ipv6_dns_server(child):
+    child.sendline('dns server 6 2001:4860:4860::8888')
+    child.sendline('dns server')
+    child.expect_exact('DNS server: [2001:4860:4860::8888]:53')
+    _test_dns_requests(child)
```
https://developers.google.com/speed/public-dns/docs/using

They are probably more stable than the example domain data.

miri64 replied:
Wouldn't be the first service Google shuts down because they feel like it (even though people rely on it)... but I am not blocking because of that. I would prefer a setup where a server is provided (using scapy e.g.), so at some point it _could_ be run automatically, but I won't block the PR because of that.

HendrikVE replied:
I agree that this would be nice, but for me (personally) it would be simply an overkill for what this test is for me: a by-product of playing around with lwIP ;) From the available DNS servers like Quad9 or Cloudflare I would say that I trust Google the most when it comes to "How long will this service still be available".

yarrick commented:
Can we do something to share the code (binary and testcase) from the `gnrc_sock_dns` test? It has an advanced test case already

It seems just selecting IP stack in the Makefile would be enough for that to work. Not sure if there are other tests that are set up like that.

yarrick commented:
The difference in supported IP version could be handled by checking support for the `ipv4`/`ipv6` pseudomodules.

HendrikVE commented:
I could try to merge both tests into `tests/sock_dns`, but I am not aware of a test that has both network stacks as options. Probably it's because of the fact that different network stacks lead to different excluded boards due to memory reasons. I think it could be possible to symlink the python script into `lwip_sock_dns` to reuse it. I'll have a look.

HendrikVE commented:
Does the python test script in `gnrc_sock_dns` work for you? I moved it from `tests-with-config` to `tests` so I could run `make test` and get it executed (is there a way to tell the directory?) and first it seemed like its working, but later I realized that the test evaluates `error resolving ...` as a successful DNS query. Nice! So I removed the first possibility in child.expect (line 135) and now it fails. The setup without the python script works for me though, but not the setup including the `Server` object in the python test script...

HendrikVE commented:
Also the README does not mention how to keep port 53 free. If `dnsmasq` returns with the message `dnsmasq: failed to create listening socket for port 53: Address already in use` then you can look with `sudo ss -lp "sport = :domain"` which process is running on it. Probably it's `systemd-resolved` so you stop it by running `sudo systemctl stop systemd-resolved`. Now the mentioned `dnsmasq` command should work.

HendrikVE commented:
Ok nevermind. I just came back to it and took a closer look. I was super confused by the README but now I know that the `dnsmasq` setup is for **manual** testing only. When running the test script (with `make test-with-config`) it does not care about `dnsmasq` at all. I'll add these details to the README as I find it very misleading without those.

HendrikVE commented:
What do I need to do to be able to connect my `native` RIOT instance via `tap0` to my `dnsmasq` instance? I tried to assign an IP via `sudo ip addr add 10.0.0.1/24 dev tap0` but it seems like this is not enough. `ifconfig` on RIOT shows `inet addr: 0.0.0.0 mask: 0.0.0.0 gw: 0.0.0.0` which must be wrong. I'm a bit lost here. Did someone already tried successfully IPv4 on native?

HendrikVE commented:
I'm happy with the mock module now, testing `sock_dns` via `gnrc` seems sufficient to me. I'll raise a new PR for the contributions made to `tests/gnrc_sock_dns` included in here.
